src/utils/marketsAndNetworksConfig.ts

```diff
@@ -143,6 +145,134 @@ export const isFeatureEnabled = {
+/**
+ * The provider will rotate rpcs on error.
+ * If provider rotates away from the first RPC, rotate back after a set interval to prioritize using most reliable RPC.
+ * If provider rotates through all rpcs, delay to avoid spamming rpcs with requests.
+ */
+class RotationProvider extends ethersProviders.BaseProvider {
+  readonly providers: StaticJsonRpcProvider[];
+  private currentProviderIndex = 0;
+  private lastFullRotationTimestamp = 0;
+  private firstRotationTimestamp = 0;
+  // after completing a full rotation of the RotationProvider, delay to avoid spamming rpcs with requests
+  private ROTATION_DELAY = 5000;
+  // if we rotate away from first rpc, return back after this delay
+  private FALL_FORWARD_DELAY = 60000;
+
+  constructor(urls: string[], chainId: number) {
+    super(chainId);
+    this.providers = urls.map((url) => new StaticJsonRpcProvider(url, chainId));
+  }
+
+  /**
+   * If provider rotates through all rpcs, delay to avoid spamming rpcs with requests
+   */
+  async delayRotation() {
+    const now = new Date().getTime();
+    const diff = now - this.lastFullRotationTimestamp;
+    if (diff < this.ROTATION_DELAY) {
+      await sleep(this.ROTATION_DELAY - diff);
+    }
+  }
+
+  /**
+   * If we rotate away from the first RPC, rotate back after a set interval to prioritize using most reliable RPC
+   */
+  async fallForwardRotation() {
+    const now = new Date().getTime();
+    const diff = now - this.firstRotationTimestamp;
+    if (diff < this.FALL_FORWARD_DELAY) {
+      await sleep(this.FALL_FORWARD_DELAY - diff);
+      this.currentProviderIndex = 0;
+    }
+  }
+
+  /**
+   * If rpc fails, rotate to next available and trigger rotation or fall forward delay where applicable
+   * @param prevIndex last updated index, checked to avoid having multiple active rotations
+   */
+  private async rotateUrl(prevIndex: number) {
+    // don't rotate when another rotation was already triggered
+    if (prevIndex !== this.currentProviderIndex) return;
+    // if we rotate away from the first url, switch back after FALL_FORWARD_DELAY
+    if (this.currentProviderIndex === 0) {
+      this.currentProviderIndex += 1;
+      this.firstRotationTimestamp = new Date().getTime();
+      this.fallForwardRotation();
+    } else if (this.currentProviderIndex === this.providers.length - 1) {
+      await this.delayRotation(); // wait for ROTATION_DELAY after completing a full rotation
+      this.currentProviderIndex = 0;
+      this.lastFullRotationTimestamp = new Date().getTime();
+    } else {
+      this.currentProviderIndex += 1;
+    }
+  }
+
+  async detectNetwork(): Promise<Network> {
+    const index = this.currentProviderIndex;
+    try {
+      return await this.providers[index].detectNetwork();
+    } catch (e) {
+      console.log(e.message);
+      await this.rotateUrl(index);
+      return this.detectNetwork();
+    }
+  }
+
+  // eslint-disable-next-line
+  async send(method: string, params: Array<any>): Promise<any> {
```

Comment:
Is this overriding a base class function? I'm not seeing it in base provider, I think this can go

Reply:
It's not on the `BaseProvider` class, but it is a function on the `JsonRpcProvider` class

Reply:
Ok, so we should be good to get rid of this since we're inheriting from `BaseProvider`